std/typecons.d

```diff
@@ -3909,6 +3909,335 @@ if (is (typeof(nullValue) == T))
+        auto spec = singleSpec("%s");
+        toString(app, spec);
+        return app.data;
+    }
+
+    /// ditto
+    string toString() const
+    {
+        import std.array : appender;
+        auto app = appender!string();
+        auto spec = singleSpec("%s");
+        toString(app, spec);
+        return app.data;
+    }
+
+    /// ditto
+    void toString(W)(ref W writer, scope const ref FormatSpec!char fmt)
+    if (isOutputRange!(W, char))
+    {
+        import std.range.primitives : put;
+        if (present)
+            formatValue(writer, _value.payload, fmt);
+        else
+            put(writer, "Optional.absent");
+    }
+
+    /// ditto
+    void toString(W)(ref W writer, scope const ref FormatSpec!char fmt) const
+    if (isOutputRange!(W, char))
+    {
+        import std.range.primitives : put;
+        if (present)
+            formatValue(writer, _value.payload, fmt);
+        else
+            put(writer, "Optional.absent");
+    }
+
+/**
+Check if `this` contains a value of `T`.
+
+Returns:
+    true $(B iff) `this` is present, otherwise false.
+ */
+    @property bool present() const @safe pure nothrow
+    {
+        return _present;
+    }
+
+///
+@safe unittest
+{
+    Optional!int optional;
+    assert(!optional.present);
+
+    optional = 0;
+    assert(optional.present);
+}
+
+// Issue 14940
+@safe unittest
+{
+    import std.array : appender;
+    import std.format : formattedWrite;
+
+    auto app = appender!string();
+    Optional!int a = 1;
+    formattedWrite(app, "%s", a);
+    assert(app.data == "1");
+}
+
+// Issue 19799
+@safe unittest
+{
+    import std.format : format;
+
+    const Optional!string a = const(Optional!string)();
+
+    format!"%s"(a);
+}
+
+/**
+Assigns `value` to the internally-held state. If the assignment
+succeeds, `this` becomes present.
+
+Params:
+    value = A value of type `T` to assign to this `Optional`.
+ */
+    void opAssign()(T value)
+    {
+        import std.algorithm.mutation : moveEmplace, move;
+
+        // the lifetime of the value in copy shall be managed by
+        // this Optional, so we must avoid calling its destructor.
+        auto copy = DontCallDestructorT(value);
+
+        if (_present)
+        {
+            move(copy.payload, _value.payload);
+        }
+        else
+        {
+            // trusted since payload is known to be T.init here.
+            () @trusted { moveEmplace(copy.payload, _value.payload); }();
+        }
+        _present = true;
+    }
+
+    immutable static missingValueException_ = new MissingValueException(
+        "Called `value' on absent Optional!" ~ T.stringof ~ ".");
+
+/**
+Returns the value if present. If `this` is in the absent state, `value` shall
+throw a `MissingValueException`.
+
+Returns:
+    The value held by this `Optional`.
+ */
+    @property ref inout(T) value() inout @nogc pure @safe
+    {
+        if (!_present)
+        {
+            throw missingValueException_;
```

Comment:
This should not be a recoverable error imo. It's more normal (and I think cleaner) for optional types to be more about asking for permission than asking for forgiveness. I'd suggest either an assertion or an Error as opposed to an Exception.

Here's an example showing how attempting to unwrap an empty optional value in Rust produces a panic (unrecoverable error) and accessing an Option's value is normally done via pattern matching (i.e. checking state first, and then accessing value): http://www.ameyalokare.com/rust/2017/10/23/rust-options.html

Reply:
This is the defining difference between `Nullable` and `Optional.` The entire reason why `get` can be aliased to this in `Nullable` _is_ that accessing a `Nullable` that is null is _undefined_, so from a certain perspective `Nullable!T` to `T` is an operation that "cannot fail" and is thus "safe". It is thus 100% deliberate for `value` to throw an `Exception`, because the state of `Optional but without a value` is a fully defined one and thus one that must result in ordinary control flow in all methods. (That's why `valueUnsafe` is called `value_Unsafe_`.)

Reply:
@FeepingCreature this is not what normally distinguishes an optional type from a nullable type.

> The core difference between option types and nullable types is that option types support nesting (Maybe (Maybe A) ≠ Maybe A), while nullable types do not (String?? = String?).

https://en.wikipedia.org/wiki/Option_type

Very often, the fundamental motivation for using an optional type is to not involve exceptions. Wrongly attempting to access the value of an optional type is a programmer error, in the same sense attempting to index an out-of-bounds array index is a programmer error. (In which case D produces an unrecoverable RangeError, not an exception.)